sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -51,75 +53,143 @@ public sealed class UseUriInsteadOfString : SonarDiagnosticAnalyzer
+                        !propertySymbol.IsOverride &&
+                        NameContainsUri(propertySymbol.Name))
                     {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3995, methodDeclaration.ReturnType.GetLocation()));
+                        c.ReportDiagnostic(Diagnostic.Create(rule_S3996, propertyDeclaration.Type.GetLocation()));
                     }
+                },
+                SyntaxKind.PropertyDeclaration);
 
-                    var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
-                    if (!stringUrlParams.Any() ||
-                        HasOverloadThatUsesUriTypeInPlaceOfString(methodSymbol, stringUrlParams))
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invokedMethodSymbol = c.SemanticModel.GetSymbolInfo(c.Node).Symbol as IMethodSymbol;
+                    if (invokedMethodSymbol == null || invokedMethodSymbol.IsInType(KnownType.System_Uri))
                     {
                         return;
                     }
 
-                    foreach (int paramIdx in stringUrlParams)
+                    var stringUrlParams = GetStringUrlParamIndexes(invokedMethodSymbol);
+                    var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(invokedMethodSymbol, stringUrlParams);
+                    if (stringUrlParams.Count > 0 && methodOverloads.Any())
                     {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3994,
-                            methodDeclaration.ParameterList.Parameters[paramIdx].Type.GetLocation()));
+                        c.ReportDiagnostic(Diagnostic.Create(rule_S4005, c.Node.GetLocation()));
                     }
                 },
-                SyntaxKind.MethodDeclaration);
+                SyntaxKind.InvocationExpression,
+                SyntaxKind.ObjectCreationExpression);
+        }
 
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+        private void VerifyDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var methodDeclaration = (BaseMethodDeclarationSyntax)context.Node;
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+            if (methodDeclaration == null || methodSymbol == null || methodSymbol.IsOverride)
+            {
+                return;
+            }
+
+            VerifyReturnType(context, methodDeclaration, methodSymbol);
+
+            var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
+            if (!stringUrlParams.Any())
+            {
+                return;
+            }
+
+            var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(methodSymbol, stringUrlParams);
+            if (methodOverloads.Any())
+            {
+                var methodOverloadSet = new HashSet<IMethodSymbol>(methodOverloads);
```

Comment:
This will be a bit more efficient if you create the `HashSet<T>` first and then check its `Count` or `Any()`, because the iterator in FindOverloads* will be executed one less time:
- `Any()` will try to create one item, then `new HasSet` will create all items (including the one, created by `Any`)